File: backend/api/repository/search_engine/schema.py
from pydantic import BaseModel
from typing import List, Optional

# Define Pydantic models for OpenAccessPdf and Tldr
class OpenAccessPdf(BaseModel):
    url: str
    status: str

class Tldr(BaseModel):
    model: str
    text: str

class CitationNormalizedPercentile(BaseModel):
    value: float
    is_in_top_1_percent: bool
    is_in_top_10_percent: bool

# Updated PaperData model
class PaperData(BaseModel):
    paperId: str
    title: str
    abstract: str
    venue: str
    year: str
    citationCount: int
    citation_normalized_percentile: Optional[CitationNormalizedPercentile]
    isOpenAccess: bool
    openAccessPdf: Optional[OpenAccessPdf]  # Nullable field
    fieldsOfStudy: List[str]
    tldr: Optional[Tldr]  # Nullable field
    similarity: float
    type: str

# QueryResult
class QueryResult(BaseModel):
    data: List[List[PaperData]]
    query: str
    final_answer: str
    followup_questions: List[str]

class searchDataResponse(BaseModel):
    searchData: QueryResult

# Updated RagDataProps model
class RagDataProps(BaseModel):
    renderedPapers: List[PaperData]
    create_new_chat_instance: bool

File: frontend/lib/paperdetails/schema.ts
// interfaces.ts

export interface citationNormalizedPercentileSchema {
    value: number;
    is_in_top_1_percent: boolean;
    is_in_top_10_percent: boolean;
};

export interface openAccessPdfSchema {
    url: string | undefined;
    status: string;
};

export interface authorSchema {
    authorId: string;
    name: string;
    url: string;
};

export interface CitationorReference {
    paperId: string;
    url: string;
    title: string;
    abstract: string | null;
    venue: string | null;
    year: number;
    referenceCount: number;
    citationCount: number;
    citation_normalized_percentile: citationNormalizedPercentileSchema;
    isOpenAccess: boolean;
    openAccessPdf: openAccessPdfSchema;
    fieldsOfStudy: [string];
    tldr: string | null;
    similarity: number | null;
    type: string
};

export interface PaperResponse {
    paperId: string;
    url: string;
    title: string;
    abstract: string | null;
    venue: string | null;
    year: number;
    referenceCount: number;
    citationCount: number;
    citation_normalized_percentile: citationNormalizedPercentileSchema;
    isOpenAccess: boolean;
    openAccessPdf: openAccessPdfSchema;
    fieldsOfStudy: [string];
    tldr: string | null;
    authors: authorSchema[];
    citations: CitationorReference[];
    references: CitationorReference[];
    type: string
}

export interface RelatedPapersLink {
    title: string;
    description: string;
    url: string
}

export interface AllRelatedPapersLinks {
    results: RelatedPapersLink[]
}